Updated credential providers to provide Async version for generating credentials.

ashishdhingra opened the pull request:
## Description
Updated credential providers to provide Async version for generating credentials.

**NOTE:**
Async methods don't have parameter for `CancellationToken` since the `GetCredentialsAsync()` doesn't have it.

## Motivation and Context
Issue #3626 and #3613
Existing customer PR #3627

## Testing
- Added test cases for `TestAssumeRoleAWSCredentialsAsyncGet` and `TestEC2InstanceMetadataSecureCredentialsSuccessAsync`.
- Dry-run
  - `DRY_RUN-e95a1656-57b3-44db-88d5-d14060e1e1db` completed successfully.
  - `DRY_RUN-d0d12849-b226-472a-ad45-a6156c073b3f` completed successfully.
  - `DRY_RUN-63e31fe8-dc91-4fcc-b06f-9a7fde0edfb6`  completed successfully.
  - `DRY_RUN-5d396d86-4d29-45b4-aca3-dcaca3a2a684` completed successfully (for commit https://github.com/aws/aws-sdk-net/pull/3681/commits/33a24930f91a93e14ed1bc2b2689bb57657fc97b).
  - `DRY_RUN-1a5ceffe-09c6-406f-a5ae-d49a72438f2b` completed successfully (for commit https://github.com/aws/aws-sdk-net/pull/3681/commits/e9ede6c25ed4a3e7e2ff9ea18b0ce7ccefcf7750)

References for SAML Federation:
- [How do I set up AD FS on an Amazon EC2 Windows instance to work with federation for an Amazon Cognito user pool?](https://repost.aws/knowledge-center/cognito-ad-fs-windows-instance)
- [Enabling Federation to AWS Using Windows Active Directory, ADFS, and SAML 2.0](https://aws.amazon.com/blogs/security/enabling-federation-to-aws-using-windows-active-directory-adfs-and-saml-2-0/)
This requires reserving Elastic IP address, create public hosted zone and registering domain name.

Federated SAML credentials cannot be tested via shared credentials file. I was able to step through the code and hack way around. The problem is that [EndpointDiscoveryHandler.ProcessEndpointDiscovery()](https://github.com/aws/aws-sdk-net/blob/24650b432f14333e68c549eacb612a1b8f6e2d5c/sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/EndpointDiscoveryHandler.cs#L136) uses non-async version to get credentials.

**Testing Customer's scenario:**
- Create a new role named `TestEC2ReadOnlyRole` with `AmazonEC2ReadOnlyAccess` permission policy and below trust relationship with the account:
  ```JSON
  {
      "Version": "2012-10-17",
      "Statement": [
          {
              "Effect": "Allow",
              "Principal": {
                  "AWS": "arn:aws:iam::<<ACCOUNT-ID>>:root"
              },
              "Action": "sts:AssumeRole",
              "Condition": {}
          }
      ]
  }
  ```
- In `~/.aws/config` file, setup below profile:
  ```
  [default]
  region = us-east-2
  output = json
  
  [profile testassumerole]
  source_profile = default
  role_arn = arn:aws:iam::<<ACCOUNT-ID>>:role/TestEC2ReadOnlyRole
  ```
  The main credentials for `default` profile are configured in `~/.aws/credentials`.
- While debugging via Visual Studio, set `AWS_PROFILE` environment variable to `testassumerole`.

While the `async` path for `ICoreAmazonSTS.CredentialsFromAssumeRoleAuthenticationAsync()` is triggered in `V4` branch for this PR, but below are my observations which highlight that `sync` path is triggered at some point. Comparison between V3 and V4 stack trace is also mentioned.

@normj / @dscpinheiro Prior to implementation of SRA, SDK used [CredentialsRetriever](https://github.com/aws/aws-sdk-net/blob/b719013e77bc436239d127b354d338c8b585fbc9/sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/CredentialsRetriever.cs#L30C18-L30C38) (refer stack trace in customer's issue #3626). It emitted [InvokeAsync()](https://github.com/aws/aws-sdk-net/blob/b719013e77bc436239d127b354d338c8b585fbc9/sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/CredentialsRetriever.cs#L90) versions. Is there an equivalent in SRA world? Now `EndpointDiscoveryHandler` always uses sync version at [EndpointDiscoveryHandler.ProcessEndpointDiscovery()](https://github.com/aws/aws-sdk-net/blob/24650b432f14333e68c549eacb612a1b8f6e2d5c/sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/EndpointDiscoveryHandler.cs#L136). We have removed `CredentialsRetriever` in V4 in [commit](https://github.com/aws/aws-sdk-net/commit/076301c8b37a29d66389c77939513ed2261e5bc3#diff-da3d680e21e83c003331723deb52ca6c2d2b46b3d7785f460957c04f5512fe82). Please suggest if async invocation for retrieving credentials is no longer possible in V4!

**V3 Call stack:**
While debugging it worked fine.
```C#
>	AWSSDK.Core.dll!Amazon.Runtime.RefreshingAWSCredentials.GenerateNewCredentialsAsync() Line 273	C#
 	AWSSDK.Core.dll!Amazon.Runtime.RefreshingAWSCredentials.GetCredentialsAsync() Line 154	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CredentialsRetriever.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 99	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.RetryHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 135	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CompressionHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 57	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 61	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.EC2.dll!Amazon.EC2.Internal.AmazonEC2PostMarshallHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 57	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.BaseEndpointResolver.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 43	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Marshaller.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 56	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.EC2.dll!Amazon.EC2.Internal.AmazonEC2PreMarshallHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 80	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 61	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 53	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 165	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.MetricsHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 94	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.RuntimePipeline.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 152	C#
 	AWSSDK.Core.dll!Amazon.Runtime.AmazonServiceClient.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.AmazonWebServiceRequest request, Amazon.Runtime.Internal.InvokeOptionsBase options, System.Threading.CancellationToken cancellationToken) Line 278	C#
 	AWSSDK.EC2.dll!Amazon.EC2.AmazonEC2Client.DescribeTagsAsync(Amazon.EC2.Model.DescribeTagsRequest request, System.Threading.CancellationToken cancellationToken) Line 16218	C#
 	AWSSDK.EC2.dll!Amazon.EC2.AmazonEC2Client.DescribeTagsAsync(System.Threading.CancellationToken cancellationToken) Line 16180	C#
 	V4Test.dll!Program.<Main>$(string[] args) Line 7	C#
```
**V4 Call stack:**
```C#
>	AWSSDK.Core.dll!Amazon.Runtime.AssumeRoleAWSCredentials.GenerateNewCredentials() Line 105	C#
 	AWSSDK.Core.dll!Amazon.Runtime.RefreshingAWSCredentials.GetCredentials.__GenerateCredentialsAndUpdateState|11_0() Line 166	C#
 	AWSSDK.Core.dll!Amazon.Runtime.RefreshingAWSCredentials.GetCredentials() Line 150	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.ProcessEndpointDiscovery(Amazon.Runtime.IRequestContext requestContext, bool evictCacheKey, System.Uri evictUri) Line 136	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.DiscoverEndpoints(Amazon.Runtime.IRequestContext requestContext, bool evictCacheKey) Line 112	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.PreInvoke(Amazon.Runtime.IExecutionContext executionContext) Line 101	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 72	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.ChecksumHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 52	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.RetryHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 135	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CompressionHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 56	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 61	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.EC2.dll!Amazon.EC2.Internal.AmazonEC2PostMarshallHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 57	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.BaseEndpointResolver.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 43	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.BaseAuthResolverHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 47	C#
 	[Async Call Stack]	
 	[Async] AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 61	C#
 	[Async] AWSSDK.Core.dll!Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 53	C#
 	[Async] AWSSDK.Core.dll!Amazon.Runtime.Internal.MetricsHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 93	C#
 	[Async] V4Test.dll!Program.<Main>$(string[] args) Line 7	C#
```
It ultimately calls `async` version while invoking Signer to sign the request:
```C#
>	AWSSDK.Core.dll!Amazon.Runtime.RefreshingAWSCredentials.GetCredentialsAsync() Line 183	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.SignRequestAsync(Amazon.Runtime.IRequestContext requestContext) Line 158	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.PreInvokeAsync(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 57	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 76	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.ChecksumHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 52	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.RetryHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 135	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CompressionHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 56	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 61	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.EC2.dll!Amazon.EC2.Internal.AmazonEC2PostMarshallHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 57	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.BaseEndpointResolver.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 43	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.BaseAuthResolverHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 47	C#
```
But looks like at some point(s) we make sync calls. And it might fail at the previous sync stack trace while assuming role, on platforms like Android which don't support sync calls (per customer's issue).

## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License
<!--- The SDK is released under the [Apache 2.0 license][license], so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a [Contributor License Agreement (CLA)][cla] -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

dscpinheiro commented on generator/.DevConfigs/9fa1d027-8266-430e-b4c0-a541eb733f8e.json:
```diff
@@ -0,0 +1,18 @@
+            "Updated credential providers to provide Async version for generating credentials."
+        ],
+        "type": "patch",
+        "updateMinimum": false
```
It doesn't matter here for V4 since we ship all packages in V4, but this should be `true` (it's rare that a change to Core won't require all services to be updated too).

ashishdhingra replied:
Fixed.

dscpinheiro commented on sdk/src/Core/Amazon.Runtime/Credentials/AWSCredentials.cs:
```diff
@@ -35,11 +35,9 @@ public abstract class AWSCredentials : BaseIdentity
         /// </summary>
         protected virtual void Validate() { }
 
-#if AWS_ASYNC_API
```
Why are we removing `AWS_ASYNC_API`? (By the way I think this constant is redundant since every target supports async but it's in a lot of other places too)

ashishdhingra replied:
I removed it based on discussion with @normj since it is now redundant. And based on discussion, the current scope of removal was limited to credential providers.

dscpinheiro commented on sdk/test/NetStandard/IntegrationTests/IntegrationTests/CredentialsTests.cs:
```diff
@@ -47,5 +50,38 @@ public async Task TestSessionCredentials()
+
+        [Fact]
+        [Trait(TestBase.CategoryAttribute, "Credentials")]
+        public async Task TestAssumeRoleAWSCredentialsAsyncGet()
```
FYI This test doesn't run in our build system (it skips the `RequiresIAMUser` category that's applied to the parent class).

ashishdhingra replied:
Thanks for letting me know. Do we have reasoning for excluding the test `RequiresIAMUser` category?

dscpinheiro replied:
It's not the case for the test you added, but for some other operations (such as https://github.com/aws/aws-sdk-net/blob/main/sdk/test/Services/SecurityToken/IntegrationTests/GetFederationToken.cs), it requires an IAM user (and our build system assumes a role in a separate account to run tests). 

From https://docs.aws.amazon.com/STS/latest/APIReference/API_GetFederationToken.html:
> Returns a set of temporary security credentials (consisting of an access key ID, a secret access key, and a security token) for a user. A typical use is in a proxy application that gets temporary security credentials on behalf of distributed applications inside a corporate network.
> You must call the GetFederationToken operation using the long-term security credentials of an IAM user.

ashishdhingra replied:
Removed test case in latest commit https://github.com/aws/aws-sdk-net/pull/3681/commits/e9ede6c25ed4a3e7e2ff9ea18b0ce7ccefcf7750. Will initiate a dry run.

dscpinheiro commented on sdk/src/Core/Amazon.Runtime/Credentials/RefreshingAWSCredentials.cs:
If you're updating this file I'm wondering if it makes sense to also fix https://github.com/aws/aws-sdk-net/issues/3613

ashishdhingra replied:
Discussed this with @norm and looked at the backlog item. The issue needs investigation per backlog item on why `preemptExpiryTime` parameter is not used, and if we should be using it. It's the pending item and we should discuss about it.

ashishdhingra replied:
@dscpinheiro I had implemented fix for #3613 in commit https://github.com/aws/aws-sdk-net/pull/3681/commits/33a24930f91a93e14ed1bc2b2689bb57657fc97b. @normj reviewed it. Just re-checking if you also reviewed it. :)

dscpinheiro reviewed (commented):
I took an initial pass but as discussed offline there's a lot of code that had to be duplicated and I'm not sure it's the best approach. 

For the question in the description, we removed the credential retriever pipeline handler and the signers now retrieve credentials themselves. The async method in the signer just invokes the sync version: https://github.com/aws/aws-sdk-net/blob/9d4fea34442970b1cdf2e1aff8b02d11e0e135fe/sdk/src/Core/Amazon.Runtime/Internal/Auth/AbstractAWSSigner.cs#L115-L130

ashishdhingra commented:
> I took an initial pass but as discussed offline there's a lot of code that had to be duplicated and I'm not sure it's the best approach.
> 
> For the question in the description, we removed the credential retriever pipeline handler and the signers now retrieve credentials themselves. The async method in the signer just invokes the sync version:
> 
> https://github.com/aws/aws-sdk-net/blob/9d4fea34442970b1cdf2e1aff8b02d11e0e135fe/sdk/src/Core/Amazon.Runtime/Internal/Auth/AbstractAWSSigner.cs#L115-L130

@dscpinheiro Thanks for the review. I understand the signer will call Async version based on stack trace I shared earlier:
```C#
AWSSDK.Core.dll!Amazon.Runtime.RefreshingAWSCredentials.GetCredentialsAsync() Line 183	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.SignRequestAsync(Amazon.Runtime.IRequestContext requestContext) Line 158	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.PreInvokeAsync(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 57	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 76	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.ChecksumHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 52	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
```

I'm referring to sync call in below stack trace after `AmazonEC2PostMarshallHandler.InvokeAsync()`:
```C#
AWSSDK.Core.dll!Amazon.Runtime.AssumeRoleAWSCredentials.GenerateNewCredentials() Line 105	C#
 	AWSSDK.Core.dll!Amazon.Runtime.RefreshingAWSCredentials.GetCredentials.__GenerateCredentialsAndUpdateState|11_0() Line 166	C#
 	AWSSDK.Core.dll!Amazon.Runtime.RefreshingAWSCredentials.GetCredentials() Line 150	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.ProcessEndpointDiscovery(Amazon.Runtime.IRequestContext requestContext, bool evictCacheKey, System.Uri evictUri) Line 136	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.DiscoverEndpoints(Amazon.Runtime.IRequestContext requestContext, bool evictCacheKey) Line 112	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.PreInvoke(Amazon.Runtime.IExecutionContext executionContext) Line 101	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 72	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.ChecksumHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 52	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.RetryHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 135	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CompressionHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 56	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 61	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 82	C#
 	AWSSDK.EC2.dll!Amazon.EC2.Internal.AmazonEC2PostMarshallHandler.InvokeAsync<Amazon.EC2.Model.DescribeTagsResponse>(Amazon.Runtime.IExecutionContext executionContext) Line 57	C#
```
`EndpointDiscoveryHandler` takes place of `CredentialsRetriever` in this flow and as part of SRA change, it was updated to make sync call. Please review.

ashishdhingra commented:
> For the question in the description, we removed the credential retriever pipeline handler and the signers now retrieve credentials themselves. The async method in the signer just invokes the sync version:
> 
> https://github.com/aws/aws-sdk-net/blob/9d4fea34442970b1cdf2e1aff8b02d11e0e135fe/sdk/src/Core/Amazon.Runtime/Internal/Auth/AbstractAWSSigner.cs#L115-L130

@dscpinheiro Thanks for the feedback. I had discussed this with @normj and fixed it in commit https://github.com/aws/aws-sdk-net/pull/3681/commits/e5a45f5e5775e666fd32d99f76353d4a6697a3f4. Please review.

normj commented on generator/.DevConfigs/9fa1d027-8266-430e-b4c0-a541eb733f8e.json:
```diff
@@ -0,0 +1,18 @@
+            "serviceName": "SecurityToken",
+            "type": "patch",
+            "changeLogMessages": [
+                "Added implementation for ICoreAmazonSTS_SAML/ICoreAmazonSTS.CredentialsFromSAMLAuthenticationAsync() and ICoreAmazonSTS.CredentialsFromAssumeRoleAuthenticationAsync()."
```
Make the change log something users will understand which don't directly interact with the `ICoreX` interfaces. Something like add async code paths for the SAML and Assume role credentials providers.

ashishdhingra replied:
Done.

normj commented on sdk/src/Core/Amazon.Runtime/Credentials/FederatedAWSCredentials.cs:
```diff
@@ -195,39 +255,59 @@ private CredentialsRefreshState Authenticate(ICredentials userCredential)
+            ICoreAmazonSTS coreSTSClient = GetSTSClient(region);
+
+            var samlCoreSTSClient
+#if NETSTANDARD
```
Why do we need this `#if`?

ashishdhingra replied:
Taking current code in V3 as example, the reason is definition in [Amazon.Runtime.SharedInterfaces](https://github.com/aws/aws-sdk-net/blob/c297b0ae98caf5c94ab63febde7b125b34dd4349/sdk/src/Core/Amazon.Runtime/SharedInterfaces/ICoreAmazonSTS.cs#L37). Here for `!BCL` targets, it defines a separate `ICoreAmazonSTS_SAML` interface declaring `CredentialsFromSAMLAuthentication()`. For BCL target, this is defined in `ICoreAmazonSTS`. There is a comment for doing this:
```C#
// In the NETSTANDARD flavors of the SDK ICoreAmazonSTS is declared without CredentialsFromSAMLAuthentication,
    }             // we cannot add a new method to the interface for backward compatibility concerns.
```
In V4, I guess, should we make this breaking change and do not consider backward compatibility? Please advise.

ashishdhingra replied:
@normj Removed `ICoreAmazonSTS_SAML` interface that was earlier used for backwards compatability.

normj commented on sdk/src/Core/Amazon.Runtime/Credentials/GenericContainerCredentials.cs:
```diff
@@ -110,6 +111,47 @@ protected override CredentialsRefreshState GenerateNewCredentials()
+                }
+                ;
+
+                AWSSDKUtils.Sleep(retry.Next());
```
Can you change this to `await Task.Delay(retry.Next())`? That avoids blocking the task for a sleep. As a separate things we should probably think about deprecating the `AWSSDKUtils.Sleep()` method now that are using targets that all support `Task.Delay`.

ashishdhingra replied:
Done. Changed to `await Task.Delay(retry.Next()).ConfigureAwait(false);`

normj commented on sdk/src/Core/Amazon.Util/EC2InstanceMetadata.cs:
```diff
@@ -480,6 +458,72 @@ public static string IdentityPkcs7
+        public static IDictionary<string, IAMSecurityCredentialMetadata> GetIAMSecurityCredentials()
+        {
+            var list = GetItems("/iam/security-credentials");
+            if (list == null)
```
Since the only thing different between these sync and async is the first method. Can you put the code after getting the list of credentials in a separate private method that takes in the list and then reuse the code between the methods.

ashishdhingra replied:
We are also using individual items and making another sync/async call `var json = await GetDataAsync("/iam/security-credentials/" + item).ConfigureAwait(false);`.

I have refactored code that deserializes credential data into separate method.

normj replied:
Fair point, I missed the `GetData` and `GetDataAsync` calls.

normj commented on sdk/src/Services/SecurityToken/Custom/SAML/SAMLAssertion.cs:
```diff
@@ -121,6 +122,79 @@ public SAMLImmutableCredentials GetRoleCredentials(
+                DurationSeconds = (int)duration.TotalSeconds
+            };
+
+#if NETSTANDARD
```
Do you need the `NETSTANDARD` block here? The comment below talks about calling `internal` methods but you are calling an async method?

normj commented on sdk/src/Core/Amazon.Runtime/Credentials/GenericContainerCredentials.cs:
```diff
@@ -110,6 +111,47 @@ protected override CredentialsRefreshState GenerateNewCredentials()
+                        throw new AmazonServiceException(string.Format(CultureInfo.InvariantCulture, "Unable to retrieve credentials. Message = \"{0}\"", exception.Message));
+                    }
+                }
+                ;
```
Dangling `;` that should be deleted.

ashishdhingra replied:
Fixed in latest commit https://github.com/aws/aws-sdk-net/pull/3681/commits/e9ede6c25ed4a3e7e2ff9ea18b0ce7ccefcf7750.

normj reviewed (approved):
One minor nit but otherwise looks good.

dscpinheiro commented on sdk/src/Core/Amazon.Runtime/Credentials/FederatedAWSCredentials.cs:
```diff
@@ -195,9 +255,79 @@ private CredentialsRefreshState Authenticate(ICredentials userCredential)
             ICoreAmazonSTS coreSTSClient = GlobalRuntimeDependencyRegistry.Instance.GetInstance<ICoreAmazonSTS>(ServiceClientHelpers.STS_ASSEMBLY_NAME, ServiceClientHelpers.STS_SERVICE_CLASS_NAME,
-                new CreateInstanceContext(new SecurityTokenServiceClientContext { Action = SecurityTokenServiceClientContext.ActionContext.AssumeRoleAWSCredentials, Region = region, ProxySettings = Options?.ProxySettings }));
-            if(coreSTSClient == null)
+                            new CreateInstanceContext(new SecurityTokenServiceClientContext { Action = SecurityTokenServiceClientContext.ActionContext.AssumeRoleAWSCredentials, Region = region, ProxySettings = Options?.ProxySettings }));
```
Nit: Remove extra space before `new CreateInstance...` (this happens in `sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/EndpointDiscoveryHandler.cs` too).

ashishdhingra replied:
Fixed in latest commit https://github.com/aws/aws-sdk-net/pull/3681/commits/e9ede6c25ed4a3e7e2ff9ea18b0ce7ccefcf7750. Will initiate a dry run.

dscpinheiro commented on sdk/src/Core/Amazon.Runtime/Credentials/InstanceProfileAWSCredentials.cs:
```diff
@@ -147,6 +148,90 @@ protected override CredentialsRefreshState GenerateNewCredentials()
+                if (null != _currentRefreshState)
+                {
+#pragma warning disable CS0612, CS0618 // Type or member is obsolete
+                    var newExpiryTime = AWSSDKUtils.CorrectedUtcNow + TimeSpan.FromMinutes(2);
```
I know you got this from existing code, but `AWSSDKUtils.CorrectedUtcNow` is not deprecated anymore so the `pragma warning disable` are redundant. You should remove them.

dscpinheiro commented on sdk/src/Core/Amazon.Runtime/SharedInterfaces/ICoreAmazonSTS.cs:
```diff
@@ -34,18 +32,26 @@ public interface ICoreAmazonSTS
-    /// a runtime dependency. This interface is implemented by the AmazonSecurityTokenServiceClient 
-    /// defined in the AWSSDK.SecurityToken assembly.
-    /// </summary>
-    public interface ICoreAmazonSTS_SAML
```
Question: Why did you delete `ICoreAmazonSTS_SAML` but not the `ICoreAmazonSTS_WebIdentity` interface? Would consolidating all the methods in a single interface be an option since this is targeting V4?

ashishdhingra replied:
Fixed in latest commit https://github.com/aws/aws-sdk-net/pull/3681/commits/e9ede6c25ed4a3e7e2ff9ea18b0ce7ccefcf7750. Will initiate a dry run.

dscpinheiro commented on sdk/src/Services/SecurityToken/Custom/SAML/ADFSAuthenticationHandlers.cs:
```diff
@@ -157,6 +244,31 @@ private static string QueryProvider(Uri identityProvider, IWebProxy proxySetting
 
+        private static HttpWebRequest CreateRequest(IWebProxy proxySettings, ICredentials credentials, string authenticationType, Uri uri, CookieContainer cookieContainer, string connectionGroup)
+        {
+#pragma warning disable SYSLIB0014 // Disable obsolete warning for WebRequest.Create because we can't switch to HttpClient while we still target .NET Framework 3.5
```
Is this `pragma warning` still needed? There shouldn't be a 3.5 target anymore.

ashishdhingra replied:
I got a Error warning after removing  `pragma warning`. Hence, didn't remove it for now.

dscpinheiro commented on sdk/test/NetStandard/IntegrationTests/IntegrationTests/CredentialsTests.cs:
```diff
@@ -47,5 +50,38 @@ public async Task TestSessionCredentials()
+                };
+                var role = await iam.CreateRoleAsync(createRoleRequest);
+
+                var sourceCredentials = FallbackCredentialsFactory.GetCredentials();
```
The `FallbackCredentialsFactory` is deprecated and it should not be used anymore (the replacement is `DefaultAWSCredentialsIdentityResolver` but as I mentioned in the other comment this test doesn't run in the build system so in my opinion we should remove it).

ashishdhingra replied:
Removed test case in latest commit https://github.com/aws/aws-sdk-net/pull/3681/commits/e9ede6c25ed4a3e7e2ff9ea18b0ce7ccefcf7750. Will initiate a dry run.

ashishdhingra commented on sdk/src/Core/Amazon.Runtime/Credentials/AssumeRoleAWSCredentials.cs:
```diff
@@ -91,13 +92,32 @@ public AssumeRoleAWSCredentials(AWSCredentials sourceCredentials, string roleArn
+            ICoreAmazonSTS coreSTSClient = GetSTSClient(region);
+
+            var credentials = coreSTSClient.CredentialsFromAssumeRoleAuthentication(RoleArn, RoleSessionName, Options);
+            _logger.InfoFormat("New credentials created for assume role that expire at {0}", credentials.Expiration.ToString("yyyy-MM-ddTHH:mm:ss.fffffffK", CultureInfo.InvariantCulture));
```
Fixed in latest commit https://github.com/aws/aws-sdk-net/pull/3681/commits/e9ede6c25ed4a3e7e2ff9ea18b0ce7ccefcf7750. Will initiate a dry run.
